### PR TITLE
Azure Provider Passthrough Actuator

### PR DESCRIPTION
**Pull request by juan-lee**
This change brings passthrough behavior to the azure provider. In other
words, root credentials which have contributor or greater permissions to
the openshift cluster's resource group on azure are reused for each
credentials request. The installer ensures that the root credential is
at least contributor as part of initial provisioning.

The next change in this series will introduce the credential minting
actuator.

### CODE REVIEW

File: pkg/apis/cloudcredential/v1/azure_types.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	Scope string `json:"scope"`
+}
+
+// AzureProviderStatus contains the status of the credentials request in Azure.
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type AzureProviderStatus struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Name is the service principal created in Azure for these credentials.
+	Name string `json:"name"`
```
Comment (dgoodwin) [inline pkg/apis/cloudcredential/v1/azure_types.go]: This looked slightly ambiguous when I saw it used in code, would it be worth going to ServicePrincipal or ServicePrincipalName?

Comment (juan-lee) [reply pkg/apis/cloudcredential/v1/azure_types.go]: Looking at this again, I completely agree. I also changed ClientID to AppID to match the names in our azure sdk/api.

File: pkg/azure/client.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+	minterv1 "github.com/openshift/cloud-credential-operator/pkg/apis/cloudcredential/v1"
+	"github.com/openshift/cloud-credential-operator/pkg/controller/credentialsrequest/actuator"
+	"github.com/openshift/cloud-credential-operator/pkg/controller/secretannotator"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	RootSecretNamespace = "openshift-config"
+	RootSecretName      = "azure-creds"
```
[the remark this answers is not in the document]
Comment (juan-lee) [reply pkg/azure/client.go]: @joelddiaz tipped me off. 😀

File: pkg/controller/secretannotator/secretannotator_controller.go [hunk 3]
```diff
@@ -63,6 +63,9 @@ const (
 
 	AwsAccessKeyName       = "aws_access_key_id"
 	AwsSecretAccessKeyName = "aws_secret_access_key"
+
+	AzureClientID     = "azure_client_id"
+	AzureClientSecret = "aws_secret_access_key"
```
[the remark this answers is not in the document]
Comment (juan-lee) [reply pkg/controller/secretannotator/secretannotator_controller.go]: good catch!

File: pkg/azure/actuator_test.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+		errRegexp string
+	}{
+		{validRootSecret, ""},
+		{rootSecretBadAnnotation, "invalid mode"},
+		{rootSecretNoAnnotation, "cannot proceed without cloud cred secret annotation.*"},
+	}
+
+	for _, tt := range tests {
+		t.Run("", func(t *testing.T) {
+			f := newFakeClient(t)
```
Comment (joelddiaz) [inline pkg/azure/actuator_test.go]: i see the custom fake client you made, and i'm wondering why you didn't go with the fake client you get from "sigs.k8s.io/controller-runtime/pkg/client/fake"

File: pkg/controller/controller.go [hunk 5]
```diff
@@ -52,22 +53,29 @@ func AddToManager(m manager.Manager) error {
...
 				return err
 			}
-		} else {
+		case configv1.AzurePlatformType:
+			log.Info("initializing Azure actuator")
+			a, err = azure.NewActuator(m.GetClient())
+			if err != nil {
+				return err
+			}
+		default:
```
Comment (joelddiaz) [inline pkg/controller/controller.go]: maybe update the log message below to something like "initializing no-op actuator (not a supported platform)"

File: pkg/azure/actuator_test.go (hunk 4 above)
Comment (juan-lee) [reply pkg/azure/actuator_test.go]: didn't know about that one. Thank you for pointing it out. I'll certainly use it again in the future.

File: pkg/azure/actuator_test.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+		in        corev1.Secret
+		errRegexp string
+	}{
+		{validRootSecret, ""},
+		{rootSecretBadAnnotation, "invalid mode"},
+		{rootSecretNoAnnotation, "cannot proceed without cloud cred secret annotation.*"},
+	}
+
+	for _, tt := range tests {
+		t.Run("", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (juan-lee) [reply pkg/azure/actuator_test.go]: no particular reason, just an oversight. Thanks for pointing it out.

[the remark this answers is not in the document]
Comment (joelddiaz) [reply pkg/azure/actuator_test.go]: i see you added the name field, but the t.Run() still has the empty string as the first parameter instead of tt.name

**Comment by joelddiaz** [comment]
@juan-lee i think just adding that tt.name to the t.Run() call and we're good to go

**Comment by juan-lee** [comment]
> @juan-lee i think just adding that tt.name to the t.Run() call and we're good to go

rookie mistake, added the code, but didn't use it :P thanks!

**Pull request merged**